src/library_exceptions.js

```diff
@@ -7,92 +7,196 @@
+    DESTRUCTOR_OFFSET: Runtime.POINTER_SIZE,
+    REFCOUNT_OFFSET: Runtime.POINTER_SIZE * 2,
+    CAUGHT_OFFSET: Runtime.POINTER_SIZE * 2 + 4,
+    RETHROWN_OFFSET: Runtime.POINTER_SIZE * 2 + 5,
+
+    // Should be multiple of allocation alignment.
+    SIZE: alignMemory(Runtime.POINTER_SIZE * 2 + 6, 16)
+  },
+
+  ExceptionInfo__deps: ['__ExceptionInfoAttrs'],
+  ExceptionInfo: function(excPtr) {
+    this.excPtr = excPtr;
+    this.ptr = excPtr - ___ExceptionInfoAttrs.SIZE;
+
+    this.set_type = function(type) {
+      {{{ makeSetValue('this.ptr', '___ExceptionInfoAttrs.TYPE_OFFSET', 'type', '*') }}};
+    }
+
+    this.get_type = function() {
+      return {{{ makeGetValue('this.ptr', '___ExceptionInfoAttrs.TYPE_OFFSET', '*') }}};
+    }
+
+    this.set_destructor = function(destructor) {
+      {{{ makeSetValue('this.ptr', '___ExceptionInfoAttrs.DESTRUCTOR_OFFSET', 'destructor', '*') }}};
+    }
+
+    this.get_destructor = function() {
+      return {{{ makeGetValue('this.ptr', '___ExceptionInfoAttrs.DESTRUCTOR_OFFSET', '*') }}};
+    }
+
+    this.set_refcount = function(refcount) {
+      {{{ makeSetValue('this.ptr', '___ExceptionInfoAttrs.REFCOUNT_OFFSET', 'refcount', '*') }}};
+    }
+
+    this.set_caught = function (caught) {
+      caught = caught ? 1 : 0;
+      {{{ makeSetValue('this.ptr', '___ExceptionInfoAttrs.CAUGHT_OFFSET', 'caught', 'i8') }}};
+    }
+
+    this.get_caught = function () {
+      return {{{ makeGetValue('this.ptr', '___ExceptionInfoAttrs.CAUGHT_OFFSET', 'i8') }}} != 0;
+    }
+
+    this.set_rethrown = function (rethrown) {
+      rethrown = rethrown ? 1 : 0;
+      {{{ makeSetValue('this.ptr', '___ExceptionInfoAttrs.RETHROWN_OFFSET', 'rethrown', 'i8') }}};
+    }
+
+    this.get_rethrown = function () {
+      return {{{ makeGetValue('this.ptr', '___ExceptionInfoAttrs.RETHROWN_OFFSET', 'i8') }}} != 0;
+    }
+
+    this.init = function(type, destructor) {
+      this.set_type(type);
+      this.set_destructor(destructor);
+      this.set_refcount(0);
+      this.set_caught(false);
+      this.set_rethrown(false);
+    }
+
+    this.add_ref = function() {
+#if USE_PTHREADS
+      Atomics.add(HEAP32, (this.ptr + ___ExceptionInfoAttrs.REFCOUNT_OFFSET) >> 2, 1);
+#else
+      var value = {{{ makeGetValue('this.ptr', '___ExceptionInfoAttrs.REFCOUNT_OFFSET', 'i32') }}};
+      {{{ makeSetValue('this.ptr', '___ExceptionInfoAttrs.REFCOUNT_OFFSET', 'value + 1', 'i32') }}};
 #endif
-          return ptr;
-        }
+    }
+
+    // Returns true if last reference released.
+    this.release_ref = function() {
+#if USE_PTHREADS
+      var prev = Atomics.sub(HEAP32, (this.ptr + ___ExceptionInfoAttrs.REFCOUNT_OFFSET) >> 2, 1);
+#else
+      var prev = {{{ makeGetValue('this.ptr', '___ExceptionInfoAttrs.REFCOUNT_OFFSET', 'i32') }}};
+      {{{ makeSetValue('this.ptr', '___ExceptionInfoAttrs.REFCOUNT_OFFSET', 'prev - 1', 'i32') }}};
+#endif
+      if (prev === 0) {
+        err('Exception reference counter underflow');
```

Comment:
these 3 lines can be like they were before,
```javascript
#if ASSERTIONS
    assert(prev > 0);
#endif
```
In general we shouldn't emit logging code in a normal build. In a build with assertions we can assert on things, and if we want logging, that should be behind `EXCEPTION_DEBUG`.